### PR TITLE
Stats Revamp: Followers Donut Graph

### PR DESCRIPTION
**Pull request by sla8c**
This PR adds the Followers Donut Graph to the Total Followers Details Screen

Ref #18605

Implementation
image

Design
image

To test:
1. enable stats feature flags 
2. Go to Stats -> Insights.  Click Followers Totals
3. Followers Donut Graph should appear.  Try for a few sites with different Wordpress / email / publicise counts and the donut graph should perform what's mentioned here https://github.com/wordpress-mobile/WordPress-iOS/pull/18910#issuecomment-1163283723 

## Regression Notes
1. Potential unintended areas of impact
None

4. What I did to test those areas of impact (or what existing automated tests I relied on)
None

5. What automated tests I added (or what prevented me from doing so)
None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsInsightsDetailsViewModel.swift [hunk 1]
```diff
@@ -314,6 +313,41 @@ class SiteStatsInsightsDetailsViewModel: Observable {
...
+                let chartViewModel = StatsFollowersChartViewModel(dotComFollowersCount: 88,
+                                                                  emailFollowersCount: 44,
+
+                                                                  publicizeCount: 16)
+
+                //this doesn't segment correctly
+//                let chartViewModel = StatsFollowersChartViewModel(dotComFollowersCount: 632,
+//                                                                  emailFollowersCount: 1,
+//                                                                  publicizeCount: 16)
+
```
Comment (sla8c) [inline WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsInsightsDetailsViewModel.swift]: 3 different ChartViewModel's that can be tweaked / played with.  1st one is actual data.  2nd and 3rd are fixed values

**Comment by frosty** [comment]
I've pushed some changes that I think should resolve the issues we were seeing! It's not perfect, but it works. The issue was that we have a minimum size we can display a segment due to the large rounded endcaps, but small segments will be below that size and not display correctly.

To account for this, I made a few changes:

* We now convert all segment values to percentages when the chart is configured, as this makes things more consistent and fixes some issues that cropped up when working with sites with millions of followers.
* We now check through all segments and find any that are smaller than the minimum displayable size. We'll now always display these at a set minimum size, and then we'll adjust the remaining segments to account for the extra space. So, the chart doesn't entirely match the reality of the numbers, but it's a good visualization.
* We do a couple of passes of this, because in reducing the size of the remaining segments we could inadvertently make one of _those_ too small.

Here's a few examples of it working:

|   |   |   |   |
|---|---|---|---|
| Simulator Screen Shot - iPhone 13 Pro - 2022-06-22 at 16 34 14 | Simulator Screen Shot - iPhone 13 Pro - 2022-06-22 at 16 28 53 | Simulator Screen Shot - iPhone 13 Pro - 2022-06-22 at 15 42 54 | Simulator Screen Shot - iPhone 13 Pro - 2022-06-22 at 15 42 33 |

File: WordPress/Classes/Stores/StatsInsightsStore.swift [hunk 2]
```diff
@@ -955,6 +955,17 @@ extension StatsInsightsStore {
         return state.publicizeFollowers
     }
 
+    /// TODO - refactor other functions to use this however cannot find a valid site where publicize > 0 (yet)
```
Comment (frosty) [inline WordPress/Classes/Stores/StatsInsightsStore.swift]: Do we need to do this before merging?

File: WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsInsightsDetailsViewModel.swift [hunk 3]
```diff
@@ -314,6 +314,30 @@ class SiteStatsInsightsDetailsViewModel: Observable {
...
+                    let chartView: UIView = chartViewModel.makeFollowersChartView()
+
+                    var chartRow = TopTotalsPeriodStatsRow(itemSubtitle: "",
+                            dataSubtitle: "",
+                            dataRows: followersRowData(dotComFollowersCount: dotComFollowersCount,
+                                                                             emailFollowersCount: emailFollowersCount,
+                                                                             othersCount: publicizeCount,
+                                                                             totalCount: dotComFollowersCount + emailFollowersCount + publicizeCount),
+                            statSection: StatSection.insightsFollowersWordPress,
+                            siteStatsPeriodDelegate: nil, //TODO - look at if I need to be not null
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsInsightsDetailsViewModel.swift:335]: Did you look into this?

File: WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/StatsFollowersChartViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import Foundation
+
+struct StatsFollowersChartViewModel {
+
+    public let dotComFollowersCount: Int
+    public let emailFollowersCount: Int
+    public let publicizeCount: Int
+
+    func makeFollowersChartView() -> UIView {
+        // The fdllowers chart currently shows 3 segments. If available, it will show:
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/StatsFollowersChartViewModel.swift]: Nitpick typo `fdllowers`

**Review by frosty — COMMENTED** [review]
Looks good to me, just a few small code notes. I guess you should also review the chart code I added here!

The only slight 'quirk' I've noticed (although I think it's a general stats issue rather than directly related to this feature) is that often when I tap in I see some different data in the chart first, and then the correct data 'pops' in.

File: WordPress/Classes/Stores/StatsInsightsStore.swift (hunk 2 above)
Comment (sla8c) [reply WordPress/Classes/Stores/StatsInsightsStore.swift]: Thanks for reminding me - I've committed a change for this

File: WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsInsightsDetailsViewModel.swift (hunk 3 above)
Comment (sla8c) [reply WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsInsightsDetailsViewModel.swift:335]: Still on the TODO list thx

File: WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/StatsFollowersChartViewModel.swift (hunk 4 above)
Comment (sla8c) [reply WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/StatsFollowersChartViewModel.swift]: Done!

**Comment by sla8c** [comment]
> I guess you should also review the chart code I added here!

Sorry I should have explicitly said - I did review your chart code + tested and I don't have any suggestions to make given current timeline + but it looks fine overall 🙂

**Review by frosty — APPROVED** [review]

**Pull request merged**